File: src/main/java/nieboczek/lifestolen/serializer/base/MapSerializer.java
package nieboczek.lifestolen.serializer.base;

import nieboczek.lifestolen.serializer.lang.SerializedStringBuilder;
import nieboczek.lifestolen.serializer.lang.TokenStream;
import nieboczek.lifestolen.serializer.lang.TokenType;

import java.util.HashMap;
import java.util.Map;

public class MapSerializer<K, V> extends Serializer<Map<K, V>> {
    private final Serializer<K> keySerializer;
    private final Serializer<V> valueSerializer;

    private MapSerializer(Serializer<K> keySerializer, Serializer<V> valueSerializer) {
        this.keySerializer = keySerializer;
        this.valueSerializer = valueSerializer;
    }

    public static <K, V> MapSerializer<K, V> of(Serializer<K> keySerializer, Serializer<V> valueSerializer) {
        return new MapSerializer<>(keySerializer, valueSerializer);
    }

    @Override
    public void serialize(Map<K, V> value, SerializedStringBuilder builder) {
        builder.text('{').newLine();
        builder.indent();
        for (Map.Entry<K, V> entry : value.entrySet()) {
            builder.indented();
            keySerializer.serialize(entry.getKey(), builder);
            builder.text(" = ");
            valueSerializer.serialize(entry.getValue(), builder);
            builder.text(';').newLine();
        }
        builder.unindent();
        builder.indented().text('}');
    }

    @Override
    public Map<K, V> deserialize(TokenStream stream) {
        stream.expect(TokenType.L_BRACE);
        Map<K, V> map = new HashMap<>();
        while (stream.peek().type() != TokenType.R_BRACE) {
            K key = keySerializer.deserialize(stream);
            stream.expect(TokenType.EQUAL);
            V value = valueSerializer.deserialize(stream);
            stream.expect(TokenType.SEMICOLON);
            map.put(key, value);
        }
        stream.expect(TokenType.R_BRACE);
        return map;
    }
}

File: src/main/kotlin/nieboczek/lifestolen/serializer/base/MapSerializer.kt
package nieboczek.lifestolen.serializer.base

import nieboczek.lifestolen.serializer.lang.SerializedStringBuilder
import nieboczek.lifestolen.serializer.lang.TokenStream
import nieboczek.lifestolen.serializer.lang.TokenType

class MapSerializer<K, V>(
    private val keySerializer: Serializer<K>,
    private val valueSerializer: Serializer<V>
) : Serializer<Map<K, V>>() {
    override fun serialize(value: Map<K, V>, builder: SerializedStringBuilder) {
        builder.text('{').newLine()
        builder.indent()
        for (entry in value.entries) {
            builder.indented()
            keySerializer.serialize(entry.key, builder)
            builder.text(" = ")
            valueSerializer.serialize(entry.value, builder)
            builder.text(';').newLine()
        }
        builder.unindent()
        builder.indented().text('}')
    }

    override fun deserialize(stream: TokenStream): Map<K, V> {
        stream.expect(TokenType.L_BRACE)
        val map = HashMap<K, V>()

        while (stream.continueIfNot(TokenType.R_BRACE)) {
            val key = keySerializer.deserialize(stream)
            stream.expect(TokenType.EQUAL)
            val value = valueSerializer.deserialize(stream)
            stream.expect(TokenType.SEMICOLON)
            map[key] = value
        }

        return map
    }
}
